### PR TITLE
Start additional threads only when necessary

### PR DESCRIPTION
**Pull request by wronglink**
I have several troubles with `JobQueue` and `Dispatcher` automatically start new threads during it's initialisation. I guess it's quite unexpected behaviour to have a branch of threads just by creating new `Updater`, without starting it. Instead of it, I propose the `Updater` instance should start all necessary threads on inside it's start methods: `start_polling` and `start_webhook`.

### CODE REVIEW

File: telegram/ext/dispatcher.py [hunk 1]
```diff
@@ -193,6 +192,7 @@ def start(self):
             self.logger.error(msg)
             raise TelegramError(msg)
 
+        self._init_async_threads(uuid4(), self.workers)
```
Comment (jh0ker) [inline telegram/ext/dispatcher.py]: I think I agree with this change, but I'd like @tsnoam to give it a look as well.

File: telegram/ext/jobqueue.py [hunk 2]
```diff
@@ -51,10 +49,6 @@ def __init__(self, bot, prevent_autostart=False):
         """:type: float"""
         self._running = False
```
Comment (jh0ker) [inline telegram/ext/jobqueue.py]: Same here

File: telegram/ext/jobqueue.py [hunk 3]
```diff
@@ -51,10 +49,6 @@ def __init__(self, bot, prevent_autostart=False):
...
 
-        if not prevent_autostart:
-            self.logger.debug('Auto-starting %s', self.__class__.__name__)
-            self.start()
-
     def put(self, job, next_t=None):
         """Queue a new job. If the JobQueue is not running, it will be started.
```
Comment (jh0ker) [inline telegram/ext/jobqueue.py]: This docstring is wrong (and was wrong before your PR). The job queue is never started in this method. Please remove that sentence.

File: telegram/ext/updater.py [hunk 4]
```diff
@@ -157,6 +157,7 @@ def start_polling(self,
                 self.running = True
 
                 # Create & start threads
+                self._init_thread(self.job_queue.start, "job_queue")
```
Comment (jh0ker) [inline telegram/ext/updater.py]: Here you can simply call `self.job_queue.start()`, I believe. The method is already spawning a new thread.

File: telegram/ext/updater.py [hunk 5]
```diff
@@ -208,6 +209,7 @@ def start_webhook(self,
...
```
Comment (jh0ker) [inline telegram/ext/updater.py]: See above comment

**Comment by jh0ker** [comment]
Hey, thanks! I did a short code review. As mentioned in the review, I am hereby pinging @tsnoam to check out the code as well ;)

File: telegram/ext/updater.py (hunk 4 above)
Comment (tsnoam) [inline telegram/ext/updater.py]: +1 @jh0ker  comment here.
@wronglink can you please fix?

**Comment by wronglink** [comment]
@jh0ker @tsnoam I've updated PR and added deprecation warning about `prevent_autostart`, so I guess it could be merged in 5.2.

**Comment by tsnoam** [comment]
for some reason, i couldn't see the sources with decreased coverage. i'm attempting rebuild with travis, maybe it'll help.

**Comment by wronglink** [comment]
@tsnoam I've merged master branch, so forced TravisCI to recalculate coverage. Now files with decreased coverage are showed on coveralls.

**Comment by tsnoam** [comment]
@wronglink 
thanks for that.
seems like the decrease in coverage is due to removed lines, changing stats in 0.1%. so i'm going merge now.
thanks again for your contribution.

**Pull request merged**